features/dd-sdk-android-flags/src/main/kotlin/com/datadog/android/flags/featureflags/internal/DatadogFlagsClient.kt

```diff
@@ -73,144 +75,356 @@ internal class DatadogFlagsClient(
+     * extracting the variant, reason, and metadata fields according to the OpenFeature spec.
+     *
+     * @param T The type of the flag value.
+     * @param precomputedFlag The precomputed flag data from the repository.
+     * @param value The typed value after successful parsing.
+     * @return [ResolutionDetails] with the resolved value and associated metadata.
+     */
+    private fun <T> createSuccessResolution(precomputedFlag: PrecomputedFlag, value: T): ResolutionDetails<T> =
+        ResolutionDetails(
+            value = value,
+            variant = precomputedFlag.variationKey.takeIf { it.isNotBlank() },
+            reason = precomputedFlag.reason.takeIf { it.isNotBlank() },
+            errorCode = null,
+            errorMessage = null,
+            flagMetadata = extractMetadata(precomputedFlag.extraLogging)
         )
-    }
 
-    private companion object {
-        private const val ERROR_NO_EVALUATION_CONTEXT =
-            "No evaluation context found, exposures cannot be sent to the flags backend. " +
-                "Please call client.setContext with a valid context."
-        private const val ERROR_PARSING_JSON = "Failed to parse JSON for key: %s"
+    /**
+     * Creates an error resolution details with the default value.
+     *
+     * This helper method creates a ResolutionDetails structure for error scenarios,
+     * populating the error code, error message, and reason according to OpenFeature spec.
+     *
+     * @param T The type of the flag value.
+     * @param flagKey The flag key that was attempted to be resolved.
+     * @param defaultValue The default value to return.
+     * @param errorCode The error code indicating the type of failure.
+     * @param errorMessage A human-readable description of the error.
+     * @return [ResolutionDetails] with the default value and error information.
+     */
+    private fun <T> createErrorResolution(
+        flagKey: String,
+        defaultValue: T,
+        errorCode: ErrorCode,
+        errorMessage: String
+    ): ResolutionDetails<T> = ResolutionDetails(
+        value = defaultValue,
+        variant = null,
+        reason = "ERROR",
+        errorCode = errorCode,
+        errorMessage = "Flag '$flagKey': $errorMessage",
+        flagMetadata = null
+    )
+
+    /**
+     * Extracts metadata from a JSONObject into an immutable Map.
+     *
+     * This helper method converts the extraLogging JSONObject from PrecomputedFlag
+     * into a Map<String, Any> suitable for the ResolutionDetails metadata field.
+     * Only string, number, and boolean values are included; other types are omitted.
+     * The returned map is immutable to prevent external modifications.
+     *
+     * @param extraLogging The JSONObject containing additional flag metadata.
+     * @return An immutable map of metadata, or null if the JSONObject is empty.
+     */
+    private fun extractMetadata(extraLogging: JSONObject): Map<String, Any>? {
```

Comment:
do we need to explicitly make return type nullable? Cannot we return empty collection if `JSONObject` is empty?

Reply:
good idea. there is no reason to return null